test/e2e/storage/external/external.go

```diff
@@ -0,0 +1,214 @@
+limitations under the License.
+*/
+
+package external
+
+import (
+	"flag"
+	"io/ioutil"
+
+	"github.com/pkg/errors"
+
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	// "k8s.io/klog"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/testsuites"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// List of testSuites to be executed for each external driver.
+var csiTestSuites = []func() testsuites.TestSuite{
+	testsuites.InitVolumesTestSuite,
+	testsuites.InitVolumeIOTestSuite,
+	testsuites.InitVolumeModeTestSuite,
+	testsuites.InitSubPathTestSuite,
+	testsuites.InitProvisioningTestSuite,
+}
+
+func init() {
+	flag.Var(testDriverParameter{}, "storage.testdriver", "name of a .yaml or .json file that defines a driver for storage testing, can be used more than once")
+}
+
+type testDriverParameter struct {
+}
+
+func (t testDriverParameter) String() string {
+	return "<.yaml or .json file>"
+}
+
+func (t testDriverParameter) Set(filename string) error {
+	driver, err := t.loadDriverDefinition(filename)
+	if err != nil {
+		return err
+	}
+	if driver.DriverInfo.Name == "" {
+		return errors.Errorf("%q: DriverInfo.Name not set", filename)
+	}
+
+	// klog.Infof("Test driver definition from %q: %+v", filename, driver)
+	description := "External Storage " + testsuites.GetDriverNameWithFeatureTags(driver)
+	// klog.Infof("Adding new storage test: %s", description)
+	Describe(description, func() {
+		testsuites.DefineTestSuite(driver, csiTestSuites)
+	})
+
+	return nil
+}
+
+func (t testDriverParameter) loadDriverDefinition(filename string) (*driverDefinition, error) {
+	if filename == "" {
+		return nil, errors.New("missing file name")
+	}
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	// Some reasonable defaults follow.
+	driver := &driverDefinition{
+		DriverInfo: testsuites.DriverInfo{
+			SupportedFsType: sets.NewString(
+				"", // Default fsType
+			),
+		},
+		ClaimSize: "5Gi",
+	}
+	// TODO: strict checking of the file content once https://github.com/kubernetes/kubernetes/pull/71589
+	// or something similar is merged.
+	if err := runtime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), data, driver); err != nil {
+		return nil, errors.Wrap(err, filename)
+	}
+	return driver, nil
+}
+
+var _ testsuites.TestDriver = &driverDefinition{}
+
+// We have to implement the interface because dynamic PV may or may
+// not be supported. driverDefinition.SkipUnsupportedTest checks that
+// based on the actual driver definition.
+var _ testsuites.DynamicPVTestDriver = &driverDefinition{}
+
+// Needed only for deserialization which isn't done, therefore the
+// functions below can be empty stubs.
+var _ runtime.Object = &driverDefinition{}
```

Comment:
is its "*only* needed for deserialization which isnt done" then why do we need it at all?

Reply:
`driverDefinition` must implement the interface, otherwise it cannot be passed to `runtime.DecodeInto`. But `runtime.DecodeInto` then never calls these methods.

If there is a better method for parsing a .yaml or .json file into a normal struct, then I am all ears.